parsers/CA_SK.py

```diff
@@ -1,142 +1,107 @@
+from zoneinfo import ZoneInfo
 
-from pytz import timezone
-from requests import Response, Session
+from requests import Session
 
+from electricitymap.contrib.config import ZoneKey
+from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
+from electricitymap.contrib.lib.models.events import ProductionMix, TotalConsumption
 from parsers.lib.exceptions import ParserException
 
-TIMEZONE = timezone("America/Regina")
-
-# URLs for the different endpoints.
+TIMEZONE = ZoneInfo("America/Regina")
 PRODUCTION_URL = (
-    "https://www.saskpower.com/ignitionapi/PowerUseDashboard/GetPowerUseDashboardData"
+    "https://www.saskpower.com/ignitionapi/PowerUseDashboard"
+    "/GetPowerUseDashboardData"
 )
 CONSUMPTION_URL = "https://www.saskpower.com/ignitionapi/Content/GetNetLoad"
-
 PRODUCTION_MAPPING = {
     "Hydro": "hydro",
     "Wind": "wind",
     "Solar": "solar",
     "Natural Gas": "gas",
     "Coal": "coal",
-    "Other": "unknown",  # This is internal consumption, losses, heat recovery facilities and small independent power producers.
+    # "Other" represents internal consumption, losses, heat recovery facilities
+    # and small independent power producers.
+    "Other": "unknown",
 }
-
-USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36"
-
-
-def validate_zone_key(zone_key: str) -> None:
-    if zone_key != "CA-SK":
-        raise ParserException(
-            "CA_SK.py",
-            f"CA_SK.py is not designed to parse zone_key: {zone_key}.",
-            zone_key,
-        )
+USER_AGENT = (
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
+    "(KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36"
+)
 
 
-def validate_no_datetime(target_datetime: datetime | None, zone_key) -> None:
+def _request(
+    session: Session,
+    target_datetime: datetime | None,
+    url: str,
+    zone_key: ZoneKey,
+) -> dict | str:
+    # The source does not offer historical data, so bail out if it's requested.
     if target_datetime:
+        raise ParserException("CA_SK.py", "Unable to fetch historical data", zone_key)
+    # The zone key must be "CA-SK"; bail out otherwise.
+    if zone_key != "CA-SK":
+        raise ParserException("CA_SK.py", f"Cannot parse zone '{zone_key}'", zone_key)
+    # Mimic a user browser in the headers or the API will respond with a 403.
+    response = session.get(url, headers={"user-agent": USER_AGENT})
+    if not response.ok:
         raise ParserException(
             "CA_SK.py",
-            "This parser is unable to fetch historical data.",
+            f"Request to {url} failed. Response Code: {response.status_code}\n"
+            f"Error:\n{response.text}",
             zone_key,
         )
+    return response.json()
 
 
 def fetch_production(
-    zone_key: str = "CA-SK",
-    session: Session | None = None,
+    zone_key: ZoneKey = ZoneKey("CA-SK"),
+    session: Session = Session(),
```

Comment:
Yes, I brought up that problem in an [earlier comment](https://github.com/electricitymaps/electricitymaps-contrib/pull/6013#discussion_r1361358339) :wink:. I just changed it to agree with the documentation/examples in case it really was the desired behaviour; since it's not, I'll revert that commit. I guess the documentation should be updated accordingly. right?

Reply:
It should 😅.